In [1]:
import pandas as pd
import numpy as np
import os
from datetime import datetime

In [2]:
path='Files used in current analysis/Linkedin_Job_Details_26-Sep-2023-to-26-Nov-2023_15875-jobs_685-pages.csv'

In [3]:
df=pd.read_csv(path)

In [5]:
df

Job Title  \
0                              LinkedIn Data Extractor   
1                                 It Management Intern   
2                               SQL Database Developer   
3                      Business Intelligence Developer   
4                           Web Developer - Internship   
...                                                ...   
15870  Manager – Growth Chat & SMS (Digital Marketing)   
15871                         Release Support Engineer   
15872                  Information Security Specialist   
15873        FUNCTIONAL CONSULTANT ORACLE FUSION - HCM   
15874                              Community Associate   

                                                Job Info  \
0            EcomFleet · Pakistan (Remote)  \n1 week ago   
1      Digicon Valley · Lahore, Punjab, Pakistan (On-...   
2      Murkez Technologies · Lahore, Punjab, Pakistan...   
3      DevNatives · Islamabad, Islāmābād, Pakistan (H...   
4      ItCreaters · Lahore, Punjab, Pakistan (On-site...   
...                                                  ...   
15870  Bank Alfalah Limited · Karāchi, Sindh, Pakista...   
15871           Turing · Pakistan Reposted  \n1 week ago   
15872  DPL · Islamabad, Islāmābād, Pakistan  \n1 week...   
15873  Level 3 BOS · Karachi Division, Sindh, Pakista...   
15874  The Hive · Karachi Division, Sindh, Pakistan  ...   

                                             Job Type  \
0                                           Full-time   
1                                           Full-time   
2                                           Full-time   
3                                           Full-time   
4                                           Full-time   
...                                               ...   
15870  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15871        Remote\n\n \nFull-time\n\n \nEntry level   
15872       On-site\n\n \nFull-time\n\n \nEntry level   
15873  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15874                         On-site\n\n \nFull-time   

                                            Job Link  \
0      https://www.linkedin.com/jobs/view/3727054368   
1      https://www.linkedin.com/jobs/view/3723395858   
2      https://www.linkedin.com/jobs/view/3727034665   
3      https://www.linkedin.com/jobs/view/3727029273   
4      https://www.linkedin.com/jobs/view/3727034696   
...                                              ...   
15870  https://www.linkedin.com/jobs/view/3767355025   
15871  https://www.linkedin.com/jobs/view/3766809535   
15872  https://www.linkedin.com/jobs/view/3772530977   
15873  https://www.linkedin.com/jobs/view/3771943212   
15874  https://www.linkedin.com/jobs/view/3767944844   

                                               employees  
0                                        11-50 employees  
1                                        11-50 employees  
2                                       51-200 employees  
3                                        11-50 employees  
4                    Mobeen Naqvi is hiring for this job  
...                                                  ...  
15870                   5,001-10,000 employees · Banking  
15871  1,001-5,000 employees · IT Services and IT Con...  
15872   51-200 employees · IT Services and IT Consulting  
15873  201-500 employees · Outsourcing and Offshoring...  
15874                                    11-50 employees  

[15875 rows x 5 columns]

### checking duplicates

In [6]:
len(df)

15875

In [7]:
df['Job ID']=df['Job Link'].apply(lambda x:int(x.split('/')[5]))

In [8]:
len(df['Job ID'].unique())

12010

In [9]:
len(df)-len(df['Job ID'].unique())

3865

In [10]:
df['Job ID'].value_counts()

Job ID
3720211981    13
3707088816    11
3733022066    11
3733019363    11
3605634351    10
              ..
3736603529     1
3736386954     1
3736348678     1
3735669396     1
3767944844     1
Name: count, Length: 12010, dtype: int64

In [11]:
df['Job ID'].value_counts().unique()

array([13, 11, 10,  9,  8,  7,  6,  5,  4,  3,  2,  1], dtype=int64)

In [12]:
duplicates_count=df.duplicated(subset=['Job ID']).sum()
duplicates_count

3865

In [13]:
duplicates = df['Job ID'].duplicated()
df[duplicates]

Job Title  \
603                             Odoo Developer   
781    Assistant Manager Sales/Sales Executive   
834                  Back-End Network Engineer   
840                Release Management Engineer   
842                                        NaN   
...                                        ...   
15856     Assistant Manager Talent Acquisition   
15857                        Software Engineer   
15858                    Web Backend Developer   
15860                       Investment Advisor   
15871                 Release Support Engineer   

                                                Job Info  \
603    AKSiQ · Karachi Division, Sindh, Pakistan (On-...   
781    Eastern Housing Pakistan · Multan, Punjab, Pak...   
834    Turing · Pakistan (Remote) Reposted  \n2 weeks...   
840    AIIT Solutions · Rawalpindi, Punjab, Pakistan ...   
842                                                  NaN   
...                                                  ...   
15856  Zameen.com · Lahore, Punjab, Pakistan  \n2 wee...   
15857    Arrivy · Lahore, Punjab, Pakistan  \n1 week ago   
15858  Turing · Pakistan  \n2 weeks ago\n\n · 11 appl...   
15860  Opal smart · Islamabad, Islāmābād, Pakistan  \...   
15871           Turing · Pakistan Reposted  \n1 week ago   

                                             Job Type  \
603                                         Full-time   
781                                         Full-time   
834                           Full-time · Entry level   
840                                         Full-time   
842                                               NaN   
...                                               ...   
15856  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15857                         On-site\n\n \nFull-time   
15858        Remote\n\n \nFull-time\n\n \nEntry level   
15860                         On-site\n\n \nFull-time   
15871        Remote\n\n \nFull-time\n\n \nEntry level   

                                            Job Link  \
603    https://www.linkedin.com/jobs/view/3727054738   
781    https://www.linkedin.com/jobs/view/3727057147   
834    https://www.linkedin.com/jobs/view/3726337380   
840    https://www.linkedin.com/jobs/view/3727054443   
842    https://www.linkedin.com/jobs/view/3727051751   
...                                              ...   
15856  https://www.linkedin.com/jobs/view/3765449305   
15857  https://www.linkedin.com/jobs/view/3765793064   
15858  https://www.linkedin.com/jobs/view/3765780780   
15860  https://www.linkedin.com/jobs/view/3766397564   
15871  https://www.linkedin.com/jobs/view/3766809535   

                                               employees      Job ID  
603                                      11-50 employees  3727054738  
781                                    201-500 employees  3727057147  
834    1,001-5,000 employees · IT Services and IT Con...  3726337380  
840                                      11-50 employees  3727054443  
842                                                  NaN  3727051751  
...                                                  ...         ...  
15856               5,001-10,000 employees · Real Estate  3765449305  
15857                                   51-200 employees  3765793064  
15858  1,001-5,000 employees · IT Services and IT Con...  3765780780  
15860      Areesha Rashid Qureshi is hiring for this job  3766397564  
15871  1,001-5,000 employees · IT Services and IT Con...  3766809535  

[3865 rows x 6 columns]

### drop duplicates

In [14]:
df = df.drop_duplicates(subset=['Job ID'])

In [15]:
df.shape

(12010, 6)

### checking null values

In [17]:
df.isnull().sum()

Job Title    258
Job Info     258
Job Type     272
Job Link       0
employees    273
Job ID         0
dtype: int64

In [23]:
null_rows = df[df.isnull().any(axis=1)]
null_rows

Job Title Job Info Job Type  \
110         NaN      NaN      NaN   
125         NaN      NaN      NaN   
417         NaN      NaN      NaN   
420         NaN      NaN      NaN   
456         NaN      NaN      NaN   
...         ...      ...      ...   
15742       NaN      NaN      NaN   
15767       NaN      NaN      NaN   
15769       NaN      NaN      NaN   
15787       NaN      NaN      NaN   
15793       NaN      NaN      NaN   

                                                Job Link employees      Job ID  
110        https://www.linkedin.com/jobs/view/3716079439       NaN  3716079439  
125        https://www.linkedin.com/jobs/view/3728144603       NaN  3728144603  
417    https://www.linkedin.com/jobs/view/3728021212/...       NaN  3728021212  
420    https://www.linkedin.com/jobs/view/3727051751/...       NaN  3727051751  
456    https://www.linkedin.com/jobs/view/3726689376/...       NaN  3726689376  
...                                                  ...       ...         ...  
15742      https://www.linkedin.com/jobs/view/3772771813       NaN  3772771813  
15767      https://www.linkedin.com/jobs/view/3767693296       NaN  3767693296  
15769      https://www.linkedin.com/jobs/view/3766808575       NaN  3766808575  
15787      https://www.linkedin.com/jobs/view/3772859487       NaN  3772859487  
15793      https://www.linkedin.com/jobs/view/3766806694       NaN  3766806694  

[273 rows x 6 columns]

In [24]:
len(null_rows)

273

### dealing Null values

In [25]:
df= df.dropna()
df

Job Title  \
0                              LinkedIn Data Extractor   
1                                 It Management Intern   
2                               SQL Database Developer   
3                      Business Intelligence Developer   
4                           Web Developer - Internship   
...                                                ...   
15869                                  Unity Developer   
15870  Manager – Growth Chat & SMS (Digital Marketing)   
15872                  Information Security Specialist   
15873        FUNCTIONAL CONSULTANT ORACLE FUSION - HCM   
15874                              Community Associate   

                                                Job Info  \
0            EcomFleet · Pakistan (Remote)  \n1 week ago   
1      Digicon Valley · Lahore, Punjab, Pakistan (On-...   
2      Murkez Technologies · Lahore, Punjab, Pakistan...   
3      DevNatives · Islamabad, Islāmābād, Pakistan (H...   
4      ItCreaters · Lahore, Punjab, Pakistan (On-site...   
...                                                  ...   
15869  Codiea.io · Bahawalpur, Punjab, Pakistan  \n1 ...   
15870  Bank Alfalah Limited · Karāchi, Sindh, Pakista...   
15872  DPL · Islamabad, Islāmābād, Pakistan  \n1 week...   
15873  Level 3 BOS · Karachi Division, Sindh, Pakista...   
15874  The Hive · Karachi Division, Sindh, Pakistan  ...   

                                             Job Type  \
0                                           Full-time   
1                                           Full-time   
2                                           Full-time   
3                                           Full-time   
4                                           Full-time   
...                                               ...   
15869                         On-site\n\n \nFull-time   
15870  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15872       On-site\n\n \nFull-time\n\n \nEntry level   
15873  On-site\n\n \nFull-time\n\n \nMid-Senior level   
15874                         On-site\n\n \nFull-time   

                                            Job Link  \
0      https://www.linkedin.com/jobs/view/3727054368   
1      https://www.linkedin.com/jobs/view/3723395858   
2      https://www.linkedin.com/jobs/view/3727034665   
3      https://www.linkedin.com/jobs/view/3727029273   
4      https://www.linkedin.com/jobs/view/3727034696   
...                                              ...   
15869  https://www.linkedin.com/jobs/view/3767951886   
15870  https://www.linkedin.com/jobs/view/3767355025   
15872  https://www.linkedin.com/jobs/view/3772530977   
15873  https://www.linkedin.com/jobs/view/3771943212   
15874  https://www.linkedin.com/jobs/view/3767944844   

                                               employees      Job ID  
0                                        11-50 employees  3727054368  
1                                        11-50 employees  3723395858  
2                                       51-200 employees  3727034665  
3                                        11-50 employees  3727029273  
4                    Mobeen Naqvi is hiring for this job  3727034696  
...                                                  ...         ...  
15869                                     1-10 employees  3767951886  
15870                   5,001-10,000 employees · Banking  3767355025  
15872   51-200 employees · IT Services and IT Consulting  3772530977  
15873  201-500 employees · Outsourcing and Offshoring...  3771943212  
15874                                    11-50 employees  3767944844  

[11737 rows x 6 columns]

In [26]:
12010-len(null_rows)

11737

### exploring values in features

In [27]:
df['Job Title'].unique()	

array(['LinkedIn Data Extractor', 'It Management Intern',
       'SQL Database Developer', ...,
       'Manager – Growth Chat & SMS (Digital Marketing)',
       'Information Security Specialist',
       'FUNCTIONAL CONSULTANT ORACLE FUSION - HCM'], dtype=object)

In [28]:
len(df['Job Title'].unique())

6442

In [29]:
df['Job Title'].value_counts()[:30]

Job Title
Sales Executive                          142
Business Development Executive           135
Sales Specialist                          76
Business Development Manager              71
Business Development Specialist           63
Graphic Designer                          58
Web Back-End Developer                    57
Software Engineer                         56
Customer Service Representative           55
Salesperson                               50
Full Stack Engineer                       47
Senior Software Engineer                  45
Sales And Marketing Specialist            40
Wordpress Developer                       38
Call Center Representative                37
PHP Developer                             36
Customer Sales Representative             35
Sales Development Representative          35
Search Engine Optimization Specialist     33
Dotnet Developer                          33
Digital Marketing Specialist              32
Business Development Officer              32


In [30]:
df['Job Info'].unique()

array(['EcomFleet · Pakistan (Remote)  \n1 week ago',
       'Digicon Valley · Lahore, Punjab, Pakistan (On-site)  \n2 weeks ago',
       'Murkez Technologies · Lahore, Punjab, Pakistan (On-site)  \n1 week ago',
       ...,
       'DPL · Islamabad, Islāmābād, Pakistan  \n1 week ago\n\n · 83 applicants',
       'Level 3 BOS · Karachi Division, Sindh, Pakistan  \n1 week ago\n\n · 19 applicants',
       'The Hive · Karachi Division, Sindh, Pakistan  \n1 week ago'],
      dtype=object)

In [31]:
len(df['Job Info'].unique())

10209

In [32]:
df['Job Type'].unique()

array(['Full-time', 'Full-time · Mid-Senior level',
       'Internship · Internship', 'Full-time · Associate', 'Internship',
       'Full-time · Entry level', 'Part-time', 'Full-time · Internship',
       'Contract · Mid-Senior level', 'Contract', 'Contract · Associate',
       '$150/month - $300/month · Full-time', 'Full-time · Director',
       'Part-time · Associate', 'Mr Ahsan is hiring for this job',
       'Part-time · Executive', 'Contract · Executive',
       'Part-time · Mid-Senior level', 'Contract · Entry level',
       'Temporary · Mid-Senior level', 'Executive',
       'Full-time · Executive', 'Sean Sato is hiring for this job',
       'Part-time · Internship', 'Internship · Mid-Senior level',
       '$12,000/yr - $15,000/yr · Full-time · Mid-Senior level',
       '$200,000/yr · Full-time', 'Part-time · Entry level',
       'dr.ry. L. is hiring for this job',
       '10,001+ employees · Tobacco Manufacturing', 'Volunteer',
       'On-site\n\n \nFull-time\n\n \nMid-Senior l

In [33]:
len(df['Job Type'].unique())

129

In [34]:
df['employees']

0                                          11-50 employees
1                                          11-50 employees
2                                         51-200 employees
3                                          11-50 employees
4                      Mobeen Naqvi is hiring for this job
                               ...                        
15869                                       1-10 employees
15870                     5,001-10,000 employees · Banking
15872     51-200 employees · IT Services and IT Consulting
15873    201-500 employees · Outsourcing and Offshoring...
15874                                      11-50 employees
Name: employees, Length: 11737, dtype: object

In [35]:
df['employees'].unique()[:10]

array(['11-50 employees', '51-200 employees',
       'Mobeen Naqvi is hiring for this job',
       '1,001-5,000 employees · Software Development',
       '501-1,000 employees · IT Services and IT Consulting',
       '501-1,000 employees · Telecommunications',
       '1,001-5,000 employees · Banking', '201-500 employees',
       '10,001+ employees', '51-200 employees · Human Resources Services'],
      dtype=object)

In [36]:
len(df['employees'].unique())

975

### extracting company feature

In [37]:
df.loc[:,'company']=df['Job Info'].apply(lambda x:x.split('·')[0])

In [38]:
df['company'].unique()

array(['EcomFleet ', 'Digicon Valley ', 'Murkez Technologies ', ...,
       'EComia Global ', 'BUZZMAX.CA ', 'Codiea.io '], dtype=object)

In [39]:
len(df['company'].unique())

4435

In [40]:
df['company'].value_counts()

company
Crossover                                   848
Zameen.com                                  252
Turing                                      223
People (Professional Employers Pvt Ltd)     143
Zones IT Solutions                          116
                                           ... 
Smart IT Solutions                            1
Al Falah Chemicals                            1
Office Automation Group                       1
Jana.Pk - Travel & Tour Services              1
Codiea.io                                     1
Name: count, Length: 4435, dtype: int64

In [41]:
df['company'].value_counts()[:30]

company
Crossover                                         848
Zameen.com                                        252
Turing                                            223
People (Professional Employers Pvt Ltd)           143
Zones IT Solutions                                116
Afiniti                                           104
Jazz                                              101
Times Consultant (Pvt.) Ltd.                       88
Daraz                                              79
Telenor                                            70
HR Ways - Hiring Tech Talent                       67
The Indus Hospital                                 66
CureMD                                             66
Motive                                             65
PTCL.Official                                      59
Kamayi                                             57
BAT                                                56
K-Electric                                         49
Dubizzle Labs       

### extracting location

In [42]:
df.loc[:,'location']=df['Job Info'].apply(lambda x:x.split('·')[1].split()[0].strip(','))

In [43]:
df['location']

0          Pakistan
1            Lahore
2            Lahore
3         Islamabad
4            Lahore
            ...    
15869    Bahawalpur
15870       Karāchi
15872     Islamabad
15873       Karachi
15874       Karachi
Name: location, Length: 11737, dtype: object

In [44]:
df['location'].unique()

array(['Pakistan', 'Lahore', 'Islamabad', 'Karāchi', 'Peshawar',
       'Islāmābād', 'Rawalpindi', 'Multan', 'Karachi', 'Hyderabad',
       'Mirpur', 'Kharian', 'Sindh', 'Faisalabad', 'Jhelum', 'Gujrāt',
       'Korangi', 'Bahawalpur', 'Gulshan', 'Gujranwala', 'Sukkur',
       'Punjab', 'Shekhūpura', 'Sargodha', 'Wazirabad', 'Mānsehra',
       'Sialkot', 'Chakwāl', 'Muzaffargarh', 'Nawabshah', 'Mardān',
       'Gujrat', 'Dera', 'Sāhīwāl', 'Jhang', 'Sahiwal', 'Quetta',
       'Vienna', 'Rahimyar', 'Milan', 'Lodhran', 'South', 'Zhob',
       'Jamshoro', 'Khyber', 'New', 'Rakhni', 'Taxila', 'APAC', 'Azad',
       'Toba', 'Thatta', 'Jarānwāla', 'Mīrpur', 'Larkana', 'Pasrur',
       'Sarai', 'Panjgur', 'Baloch', 'Attock', 'Rahim', 'Ratodero', 'APJ',
       'Chakwal', 'İzmir', 'Miān', 'Battagram', 'Harīpur', 'Naushahro',
       'Khuzdar', 'Kasur', 'Abbottabad', 'Malir', 'Vehari', 'Kahuta',
       'Badin', 'Thal', 'Landhi', 'Shikarpur', 'Bhakkar', 'Bahrain',
       'Metro', 'Kohāt', 'Okāra', 

In [46]:
len(df['location'].unique())

149

In [47]:
df['location'].value_counts()

location
Lahore          3842
Karāchi         2274
Islamabad       1725
Pakistan         900
Karachi          790
                ... 
Sibi               1
Bin                1
Shergarh           1
Bahawalnagar       1
Sheikhupura        1
Name: count, Length: 149, dtype: int64

In [49]:
df['location'].value_counts()[:30]

location
Lahore          3842
Karāchi         2274
Islamabad       1725
Pakistan         900
Karachi          790
Rawalpindi       551
Islāmābād        309
Faisalabad       202
Multan           124
Peshawar         105
Punjab            84
Hyderabad         69
Sialkot           60
APAC              51
Gujranwala        49
APJ               38
Quetta            37
New               34
Sindh             32
South             30
Bahawalpur        29
Muzaffargarh      28
Jhelum            22
Gujrāt            21
Sargodha          15
Sukkur            14
Rahim             13
Jamshoro          12
Umarkot           11
Sahiwal           11
Name: count, dtype: int64

In [51]:
df.loc[:,'location']=df['Job Info'].apply(lambda x:x.split('·')[1].split('(')[0])

In [52]:
df['location']

0                                                Pakistan 
1                                Lahore, Punjab, Pakistan 
2                                Lahore, Punjab, Pakistan 
3                          Islamabad, Islāmābād, Pakistan 
4                                Lahore, Punjab, Pakistan 
                               ...                        
15869           Bahawalpur, Punjab, Pakistan  \n1 week ago
15870               Karāchi, Sindh, Pakistan  \n1 week ago
15872     Islamabad, Islāmābād, Pakistan  \n1 week ago\...
15873     Karachi Division, Sindh, Pakistan  \n1 week a...
15874      Karachi Division, Sindh, Pakistan  \n1 week ago
Name: location, Length: 11737, dtype: object

In [53]:
df['location'].unique()

array([' Pakistan ', ' Lahore, Punjab, Pakistan ',
       ' Islamabad, Islāmābād, Pakistan ', ...,
       ' Sakrand, Sindh, Pakistan  \n1 week ago\n\n ',
       ' Sheikhupura District, Punjab, Pakistan  \n1 week ago\n\n ',
       ' Islamabad, Punjab, Pakistan  \n2 weeks ago\n\n '], dtype=object)

In [54]:
len(df['location'].unique())

1163

In [55]:
df['location'].value_counts()

location
 Lahore, Punjab, Pakistan                                  1052
 Karāchi, Sindh, Pakistan                                   624
 Islamabad, Islāmābād, Pakistan                             518
 Lahore, Punjab, Pakistan  \n2 weeks ago                    326
 Karachi Division, Sindh, Pakistan                          242
                                                           ... 
 Gulshan Town, Sindh, Pakistan  \n7 hours ago\n\n             1
 Korangi District, Sindh, Pakistan  \n10 hours ago\n\n        1
 Jhelum, Punjab, Pakistan  \n23 hours ago\n\n                 1
 Sāhīwāl, Punjab, Pakistan  \n9 hours ago\n\n                 1
 Islamabad, Punjab, Pakistan  \n2 weeks ago\n\n               1
Name: count, Length: 1163, dtype: int64

In [56]:
df['location'].value_counts()[:30]

location
 Lahore, Punjab, Pakistan                              1052
 Karāchi, Sindh, Pakistan                               624
 Islamabad, Islāmābād, Pakistan                         518
 Lahore, Punjab, Pakistan  \n2 weeks ago                326
 Karachi Division, Sindh, Pakistan                      242
 Pakistan                                               236
 Lahore, Punjab, Pakistan  \n2 weeks ago\n\n            199
 Lahore, Punjab, Pakistan  \n1 week ago                 194
 Lahore, Punjab, Pakistan  \n1 week ago\n\n             177
 Karāchi, Sindh, Pakistan  \n1 week ago\n\n             145
 Lahore, Punjab, Pakistan  \n1 month ago                138
 Karāchi, Sindh, Pakistan  \n2 weeks ago                138
 Rawalpindi, Punjab, Pakistan                           127
 Karāchi, Sindh, Pakistan  \n2 weeks ago\n\n            127
 Islamabad, Islāmābād, Pakistan  \n2 weeks ago\n\n      126
 Pakistan  \n2 weeks ago\n\n                            124
 Islamabad, Islāmābād, Pakistan

### extracting remote/onsite

In [57]:
remote_onsite_list=[]
for info in df['Job Info']:
    try:
        remote_onsite_list.append(info.split('·')[1].split('(')[1].split(')')[0])
    except IndexError:
        remote_onsite_list.append('Not specified')
    

In [58]:
len(remote_onsite_list)

11737

In [59]:
df.loc[:,'Remote/Onsite']=remote_onsite_list

In [60]:
df['Remote/Onsite']

0               Remote
1              On-site
2              On-site
3               Hybrid
4              On-site
             ...      
15869    Not specified
15870    Not specified
15872    Not specified
15873    Not specified
15874    Not specified
Name: Remote/Onsite, Length: 11737, dtype: object

In [61]:
df['Remote/Onsite'].unique()

array(['Remote', 'On-site', 'Hybrid', 'Not specified'], dtype=object)

In [62]:
df['Remote/Onsite'].value_counts()

Remote/Onsite
Not specified    8396
On-site          2616
Remote            525
Hybrid            200
Name: count, dtype: int64

### extarctig number of applicants feature

In [63]:
n_applicant_list=[]
for info in df['Job Info']:
    try:
        n_applicant_list.append(info.split('·')[2].strip(' applicants'))
    except IndexError:
        n_applicant_list.append(np.nan)

In [64]:
n_applicant_list[:10]

[nan, nan, nan, '51', '51', nan, nan, nan, '312', '2,048']

In [65]:
df.loc[:,'n_applicant']=n_applicant_list

In [66]:
df['n_applicant'].unique()

array([nan, '51', '312', '2,048', '30', '3,159', '41', '3,008', '120',
       '685', '50', '28', '24', '734', '9', '112', '42', '10', '89', '48',
       '19', '732', '338', '18', '154', '472', '110', '32', '16', '448',
       '239', '933', '255', '15', '177', '33', '323', '207', '22', '21',
       '49', '591', '45', '29', '231', '337', '249', '123', '40', '192',
       '157', '1,386', '36', '1,186', '17', '76', '61', '38', '20', '12',
       '115', '73', '46', '315', '90', '67', '27', '23', '13', '172',
       '778', '96', '58', '216', '43', '204', '247', '380', '39', '1,768',
       '11', '310', '34', '108', '162', '853', '986', '215', '14', '94',
       '92', '168', '911', '53', '420', '31', '72', '44', '59', '132',
       '289', '35', '52', '1,379', '158', '260', '54', '250', '332',
       '649', '1,385', '153', '37', '83', '370', '287', '57', '339', '47',
       '208', '121', '56', '98', '68', '254', '259', '138', '305', '186',
       '197', '84', '795', '63', '79', '757', '166', '

In [67]:
len(df['n_applicant'].unique())

756

In [68]:
df['n_applicant'].value_counts()[:30]

n_applicant
51          287
Over 200    271
11          194
13          140
10          135
12          132
0           129
15          122
14          121
18          108
17          107
19          103
1           101
16           98
9            94
3            93
21           90
7            89
8            89
2            85
4            83
6            82
20           79
5            78
23           68
24           67
25           67
32           66
26           61
22           59
Name: count, dtype: int64

### Extracting job posting time feature

In [69]:
df.loc[:,'job posting time']=df['Job Info'].apply(lambda x:x.split('\n')[1])
df['job posting time']

0         1 week ago
1        2 weeks ago
2         1 week ago
3         1 week ago
4         1 week ago
            ...     
15869     1 week ago
15870     1 week ago
15872     1 week ago
15873     1 week ago
15874     1 week ago
Name: job posting time, Length: 11737, dtype: object

In [70]:
df['job posting time'].unique()

array(['1 week ago', '2 weeks ago', '1 day ago', '6 days ago',
       '5 days ago', '11 hours ago', '3 days ago', '3 hours ago',
       '4 days ago', '13 minutes ago', '14 minutes ago', '5 hours ago',
       '10 hours ago', '9 hours ago', '2 days ago', '13 hours ago',
       '18 hours ago', '7 hours ago', '4 hours ago', '3 weeks ago',
       '2 hours ago', '12 hours ago', '6 hours ago', '14 hours ago',
       '1 hour ago', '22 hours ago', '8 hours ago', '23 hours ago',
       '15 hours ago', '20 hours ago', '19 hours ago', '16 hours ago',
       '17 hours ago', '21 hours ago', '17 minutes ago', '48 minutes ago',
       '23 minutes ago', '49 minutes ago', '37 minutes ago',
       '42 minutes ago', '36 minutes ago', '44 minutes ago',
       '55 minutes ago', '32 minutes ago', '58 minutes ago',
       '3 minutes ago', '1 month ago', '4 weeks ago',
       'https://anitechnologies.net/ · Islamabad, Islāmābād, Pakistan Reposted  '],
      dtype=object)

In [72]:
df['job posting time'].value_counts()[:10]

job posting time
1 week ago     3578
2 weeks ago    2243
5 days ago      854
6 days ago      743
3 days ago      664
1 day ago       528
1 month ago     495
4 days ago      491
2 days ago      474
3 weeks ago     319
Name: count, dtype: int64

### Feature Eng on Job Type feature

In [73]:
df['Job Type'].unique()

array(['Full-time', 'Full-time · Mid-Senior level',
       'Internship · Internship', 'Full-time · Associate', 'Internship',
       'Full-time · Entry level', 'Part-time', 'Full-time · Internship',
       'Contract · Mid-Senior level', 'Contract', 'Contract · Associate',
       '$150/month - $300/month · Full-time', 'Full-time · Director',
       'Part-time · Associate', 'Mr Ahsan is hiring for this job',
       'Part-time · Executive', 'Contract · Executive',
       'Part-time · Mid-Senior level', 'Contract · Entry level',
       'Temporary · Mid-Senior level', 'Executive',
       'Full-time · Executive', 'Sean Sato is hiring for this job',
       'Part-time · Internship', 'Internship · Mid-Senior level',
       '$12,000/yr - $15,000/yr · Full-time · Mid-Senior level',
       '$200,000/yr · Full-time', 'Part-time · Entry level',
       'dr.ry. L. is hiring for this job',
       '10,001+ employees · Tobacco Manufacturing', 'Volunteer',
       'On-site\n\n \nFull-time\n\n \nMid-Senior l

In [74]:
len(df['Job Type'].unique())

129

## understanding options according to linkedin
****

### These are job Job type options on linkedin:
Full-time,
Part-time,
Contract,
Temporary,
Volunteer,
Internship,
Other
### These are Experience level options on linkedin:
Internship,
Entry level,
Associate,
Mid-Senior level,
Director,
Executive
### These are On-site/remote options on linkedin:
On-site,
Remote,
Hybrid

***

### counting number of elements in job type column

In [75]:
Job_Type_nElements=[]
for cell in df['Job Type']:
    if '·' in cell:
        nElements=len(cell.split('·'))
        Job_Type_nElements.append(nElements)
    elif '\n\n \n' in cell:
        nElements=len(cell.split('\n\n \n'))
        Job_Type_nElements.append(nElements)
    else:
        nElements=len(cell.split())
        Job_Type_nElements.append(nElements)        

In [86]:
set(Job_Type_nElements) 

{1, 2, 3, 4, 7}

In [84]:
df.loc[:,'Job Type nElements']=Job_Type_nElements

In [85]:
df['Job Type nElements'].value_counts()

Job Type nElements
2    5693
3    3633
1    2395
4      13
7       3
Name: count, dtype: int64

### extarcting Job type

In [87]:
inconsistant_Job_Type_Elements_1=['Executive']
inconsistant_Job_Type_Elements_1

['Executive']

In [88]:
inconsistant_Job_Type_Elements_2=['$150/month - $300/month · Full-time','$200,000/yr · Full-time','10,001+ employees · Tobacco Manufacturing','Full-time\n\n \nMid-Senior level']
inconsistant_Job_Type_Elements_2

['$150/month - $300/month · Full-time',
 '$200,000/yr · Full-time',
 '10,001+ employees · Tobacco Manufacturing',
 'Full-time\n\n \nMid-Senior level']

In [89]:
inconsistant_Job_Type_Elements_3=['$12,000/yr - $15,000/yr · Full-time · Mid-Senior level']
inconsistant_Job_Type_Elements_3

['$12,000/yr - $15,000/yr · Full-time · Mid-Senior level']

In [92]:
Job_Type_list=[]
for elmnt in df['Job Type']:
    if len(elmnt.split())==1:
        if elmnt not in inconsistant_Job_Type_Elements_1:
            Job_Type_list.append(elmnt)
        else:
            Job_Type_list.append('Not specified')
    if len(elmnt.split('·'))==2 or len(elmnt.split('\n\n \n'))==2:
        if elmnt not in inconsistant_Job_Type_Elements_2:
            if '·' in elmnt:
                Job_Type_list.append(elmnt.split('·')[0])
            elif '\n\n \n' in elmnt:
                Job_Type_list.append(elmnt.split('\n\n \n')[1])
            else:
                Job_Type_list.append('Not specified')
        else:
            if 'Full-time' in elmnt:
                Job_Type_list.append('Full-time')
            else:
                Job_Type_list.append('Not specified')
    if len(elmnt.split('·'))==3 or len(elmnt.split('\n\n \n'))==3:
        if elmnt not in inconsistant_Job_Type_Elements_3:
            Job_Type_list.append(elmnt.split('\n\n \n')[1])
        else:
            if 'Full-time' in elmnt:
                Job_Type_list.append('Full-time')
            else:
                Job_Type_list.append('Not specified')
    if len(elmnt.split())==7:
        Job_Type_list.append('Not specified')

In [93]:
len(Job_Type_list)

11738

In [98]:
Job_type_options=['Full-time','Part-time','Contract','Temporary','Volunteer','Internship','Other']
Job_type_options
# len(Job_type_options)

['Full-time',
 'Part-time',
 'Contract',
 'Temporary',
 'Volunteer',
 'Internship',
 'Other']

### extarcting Experience level from df['Job Type']

In [99]:
Job_Type_list=[]
for elmnt in df['Job Type']:
    if len(elmnt.split())==1:
        if elmnt in inconsistant_Job_Type_Elements_1:
            if elmnt=='Executive':
                Job_Type_list.append(elmnt)
        else:
            Job_Type_list.append('Not specified')
    if len(elmnt.split('·'))==2 or len(elmnt.split('\n\n \n'))==2:
        if elmnt not in inconsistant_Job_Type_Elements_2:
            if '·' in elmnt:
                Job_Type_list.append(elmnt.split('·')[1])
            # elif '\n\n \n' in elmnt:
            #     Job_Type_list.append(elmnt.split('\n\n \n')[2])
            else:
                Job_Type_list.append('Not specified')
        else:
            if 'Mid-Senior level' in elmnt:
                Job_Type_list.append('Mid-Senior level')
            else:
                Job_Type_list.append('Not specified')
                
    if len(elmnt.split('·'))==3 or len(elmnt.split('\n\n \n'))==3:
        if elmnt not in inconsistant_Job_Type_Elements_3:
            Job_Type_list.append(elmnt.split('\n\n \n')[2])
        else:
            # if 'Mid-Senior level' in inconsistant_Job: due to this problem not sepecified is in list 
            if 'Mid-Senior level' in elmnt:  
                Job_Type_list.append('Mid-Senior level')
            else:
                Job_Type_list.append('Not specified')
    if len(elmnt.split())==7:
        Job_Type_list.append('Not specified')

In [100]:
len(Job_Type_list)

11738

In [101]:
set(Job_Type_list)

{' Associate',
 ' Director',
 ' Entry level',
 ' Executive',
 ' Food and Beverage Manufacturing',
 ' IT Services and IT Consulting',
 ' Internship',
 ' Mid-Senior level',
 'Associate',
 'Contract',
 'Director',
 'Entry level',
 'Executive',
 'Full-time',
 'Internship',
 'Mid-Senior level',
 'Not specified'}

In [103]:
Experience_level_options=['Internship','Entry level','Associate','Mid-Senior level','Director','Executive']
Experience_level_options
# len(Experience_level_options)

['Internship',
 'Entry level',
 'Associate',
 'Mid-Senior level',
 'Director',
 'Executive']

### extracting On-site/remote from Job Type feature

In [104]:
Job_Type_list=[]
for elmnt in df['Job Type']:
    if len(elmnt.split())==1:
        if elmnt not in inconsistant_Job_Type_Elements_1:
            Job_Type_list.append('Not specified')
        else:
            Job_Type_list.append('Not specified')
    if len(elmnt.split('·'))==2 or len(elmnt.split('\n\n \n'))==2:
        if elmnt not in inconsistant_Job_Type_Elements_2:
            if '·' in elmnt:
                Job_Type_list.append('Not specified')
            elif '\n\n \n' in elmnt:
                Job_Type_list.append(elmnt.split('\n\n \n')[0])
            else:
                Job_Type_list.append('Not specified')
        else:
            Job_Type_list.append('Not specified')
    if len(elmnt.split('·'))==3 or len(elmnt.split('\n\n \n'))==3:
        if elmnt not in inconsistant_Job_Type_Elements_3:
            Job_Type_list.append(elmnt.split('\n\n \n')[0])
        else:
            Job_Type_list.append('Not specified')
    if len(elmnt.split())==7:
        Job_Type_list.append('Not specified')

In [105]:
len(Job_Type_list)

11738

In [108]:
On_site_remote_options=['On-site','Remote','Hybrid']
On_site_remote_options
# len(On_site_remote_options)

['On-site', 'Remote', 'Hybrid']

#### how many are specified

#### working on employees feature , Extracting no of employees and industry feature from employees feature

In [109]:
df['employees'].unique()[:10]

array(['11-50 employees', '51-200 employees',
       'Mobeen Naqvi is hiring for this job',
       '1,001-5,000 employees · Software Development',
       '501-1,000 employees · IT Services and IT Consulting',
       '501-1,000 employees · Telecommunications',
       '1,001-5,000 employees · Banking', '201-500 employees',
       '10,001+ employees', '51-200 employees · Human Resources Services'],
      dtype=object)

In [110]:
inconsistent_values_in_employees_feature = ['is hiring for this job', 
                                            'See how you compare', 
                                            'Job poster joined LinkedIn',
                                           'school alum works here',
                                            'alumni work here',
                                            'See recent hiring trends'
                                           ]

In [111]:
L1=[] #for employees 
L2=[] #for industry
for feature_element in df['employees'].iloc[:]:
    flag=0
    for value in inconsistent_values_in_employees_feature:
        if value in feature_element:
            flag=1
            
    if flag==1:
        L1.append('Not specified')
        L2.append('Not specified')
    else:
        if len(feature_element.split('·'))==1:
            L1.append(feature_element)
            L2.append('Not specified')
        elif len(feature_element.split('·'))==2:
            L1.append(feature_element.split('·')[0])
            L2.append(feature_element.split('·')[1])
        else:
            L1.append('Other')
            L2.append('Other')

In [115]:
len(L1)

11737

In [116]:
len(L2)

11737

In [117]:
df.loc[:,'Comp_N_Employees']=L1

In [118]:
df.loc[:,'Comp_Industry']=L2

In [119]:
df['Comp_N_Employees'].unique()

array(['11-50 employees', '51-200 employees', 'Not specified',
       '1,001-5,000 employees ', '501-1,000 employees ',
       '201-500 employees', '10,001+ employees', '51-200 employees ',
       '501-1,000 employees', '1,001-5,000 employees', '1-10 employees ',
       '10,001+ employees ', '1-10 employees', '5,001-10,000 employees',
       '201-500 employees ', '5,001-10,000 employees ',
       '11-50 employees ', '1 company alum works here',
       'Industrial Automation'], dtype=object)

In [120]:
len(df['Comp_N_Employees'].unique())

19

In [121]:
df['Comp_N_Employees'].value_counts()

Comp_N_Employees
11-50 employees              2258
5,001-10,000 employees       1481
51-200 employees             1407
1,001-5,000 employees        1228
10,001+ employees            1133
Not specified                 946
1-10 employees                674
201-500 employees             596
51-200 employees              464
201-500 employees             346
501-1,000 employees           340
11-50 employees               273
501-1,000 employees           189
1,001-5,000 employees         165
1-10 employees                126
5,001-10,000 employees         60
10,001+ employees              49
1 company alum works here       1
Industrial Automation           1
Name: count, dtype: int64

In [122]:
df['Comp_Industry'].unique()

array(['Not specified', ' Software Development',
       ' IT Services and IT Consulting', ' Telecommunications',
       ' Banking', ' Human Resources Services',
       ' Technology, Information and Internet',
       ' Retail Apparel and Fashion', ' International Affairs',
       ' Higher Education', ' Information Technology & Services',
       ' Financial Services', ' Packaging and Containers Manufacturing',
       ' Executive Offices', ' Food and Beverage Manufacturing',
       ' Food and Beverage Services', ' Tobacco Manufacturing',
       ' Business Consulting and Services', ' Entertainment Providers',
       ' Wholesale Building Materials',
       ' Outsourcing and Offshoring Consulting',
       ' Pharmaceutical Manufacturing', ' Motor Vehicle Manufacturing',
       ' Non-profit Organizations', ' Research Services',
       ' Advertising Services', ' Beverage Manufacturing',
       ' Outsourcing/Offshoring', ' Hospitals and Health Care',
       ' Farming', ' Oil and Gas', ' Real Est

In [123]:
len(df['Comp_Industry'].unique())

112

In [124]:
df['Comp_Industry'].value_counts()

Comp_Industry
Not specified                                6346
 IT Services and IT Consulting               1758
 Software Development                         612
 Telecommunications                           310
 Real Estate                                  290
                                             ... 
 Renewable Energy Equipment Manufacturing       1
 Writing and Editing                            1
 Furniture                                      1
 Research                                       1
 Retail Luxury Goods and Jewelry                1
Name: count, Length: 112, dtype: int64

In [125]:
df.columns

Index(['Job Title', 'Job Info', 'Job Type', 'Job Link', 'employees', 'Job ID',
       'company', 'location', 'Remote/Onsite', 'n_applicant',
       'job posting time', 'Job Type nElements', 'Comp_N_Employees',
       'Comp_Industry'],
      dtype='object')

In [127]:
len(df.columns)

14

# End Note
***due to ValueError: Length of values (11738) does not match length of index (11737) we have missed 3 features to extract from Job type features***